apps/extension/src/storage/LocalStorage.ts

```diff
@@ -0,0 +1,188 @@
+import { KVStore } from "@namada/storage";
+import * as E from "fp-ts/Either";
+import * as t from "io-ts";
+import { ExtStorage } from "./Storage";
+
+enum BridgeType {
+  IBC = "ibc",
+  Ethereum = "ethereum-bridge",
+}
+
+const Chain = t.intersection([
+  t.type({
+    alias: t.string,
+    bech32Prefix: t.string,
+    bip44: t.type({
+      coinType: t.number,
+    }),
+    bridgeType: t.array(
+      t.keyof({ [BridgeType.IBC]: null, [BridgeType.Ethereum]: null })
+    ),
+    chainId: t.string,
+    currency: t.intersection([
+      t.type({
+        symbol: t.string,
+        token: t.string,
+      }),
+      t.partial({
+        address: t.string,
+        gasPriceStep: t.type({
+          average: t.number,
+          high: t.number,
+          low: t.number,
+        }),
+      }),
+    ]),
+    extension: t.type({
+      alias: t.string,
+      id: t.keyof({ namada: null, keplr: null, metamask: null }),
+      url: t.string,
+    }),
+    id: t.keyof({
+      namada: null,
+      cosmos: null,
+      ethereum: null,
+    }),
+    rpc: t.string,
+  }),
+  t.partial({
+    ibc: t.type({
+      portId: t.string,
+    }),
+  }),
+]);
+type ChainType = t.TypeOf<typeof Chain>;
+
+const NamadaExtensionApprovedOrigins = t.array(t.string);
+type NamadaExtensionApprovedOriginsType = t.TypeOf<
+  typeof NamadaExtensionApprovedOrigins
+>;
+
+const NamadaExtensionRouterId = t.number;
+type NamadaExtensionRouterIdType = t.TypeOf<typeof NamadaExtensionRouterId>;
+
+const Tabs = t.array(t.type({ tabId: t.number, timestamp: t.number }));
+type TabsType = t.TypeOf<typeof Tabs>;
+
+type LocalStorageTypes =
+  | ChainType
+  | NamadaExtensionApprovedOriginsType
+  | NamadaExtensionRouterIdType
+  | TabsType;
+
+type LocalStorageSchemas =
+  | typeof Chain
+  | typeof NamadaExtensionApprovedOrigins
+  | typeof NamadaExtensionRouterId
+  | typeof Tabs;
+
+const keys = [
+  "chains",
+  "namadaExtensionApprovedOrigins",
+  "namadaExtensionRouterId",
+  "tabs",
+] as const;
+export type LocalStorageKeys = (typeof keys)[number];
+
+const schemasMap = new Map<LocalStorageSchemas, LocalStorageKeys>([
+  [Chain, "chains"],
+  [NamadaExtensionApprovedOrigins, "namadaExtensionApprovedOrigins"],
+  [NamadaExtensionRouterId, "namadaExtensionRouterId"],
+  [Tabs, "tabs"],
+]);
+
+export class LocalStorage extends ExtStorage {
+  constructor(provider: KVStore<LocalStorageTypes>) {
+    super(provider);
+  }
+
+  async getChain(): Promise<ChainType | undefined> {
+    const data = await this.getRaw(this.getKey(Chain));
+
+    const Schema = t.union([Chain, t.undefined]);
+    const decodedData = Schema.decode(data);
+
+    if (E.isLeft(decodedData)) {
+      throw new Error("Chain is not valid");
+    }
+
+    return decodedData.right;
+  }
+
+  async setChain(chain: ChainType): Promise<void> {
+    await this.setRaw(this.getKey(Chain), chain);
+  }
+
+  async getApprovedOrigins(): Promise<
```

Comment:
So I did it like that because of the VaultStorage. So VaultStorage has pretty generic API and I preferred to pass schema instead of key. On the other hand I wanted schemasMap to be consistent accross storages :D Not sure which approach is better. The `getKey` `!key` check is a bit of an overkill anyways as we could probably pretty safely typecast. Let me know what you think :)